src/protocol/sort/shuffle.rs

```diff
@@ -103,70 +119,121 @@ impl<'a, F: Field, S: Step + SpaceIndex> Shuffle<'a, F, S> {
+    /// Order of calling `single` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
     /// Each single shuffle requires communication between helpers to perform reshare.
     /// Infrastructure has a pre-requisite to distinguish each communication step uniquely.
-    /// For this, we have three shuffle steps one per `single_shuffle` i.e. Step1, Step2 and Step3.
+    /// For this, we have three shuffle steps one per `single` i.e. Step1, Step2 and Step3.
     /// The Shuffle object receives a step function and appends a `ShuffleStep` to form a concrete step
     /// ![Shuffle steps][shuffle]
     #[allow(dead_code)]
-    pub async fn execute<M: Mesh, G: Gateway<M, S>>(
+    pub async fn execute_shuffle<M: Mesh, G: Gateway<M, S>>(
+        &mut self,
+        ctx: &ProtocolContext<'_, G, S>,
+        permutations: &(Permutation, Permutation),
+    ) -> Result<(), BoxError>
+    where
+        F: Field,
+    {
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Shuffle, ctx, Step1, permutations)
+            .await
+            .unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Shuffle, ctx, Step2, permutations)
+            .await
+            .unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Shuffle, ctx, Step3, permutations)
+            .await
+            .unwrap();
+
+        Ok(())
+    }
+
+    #[embed_doc_image("unshuffle", "images/sort/unshuffle.png")]
+    /// Unshuffle calls `single` three times with 2 helpers shuffling the shares each time in the opposite order to shuffle.
+    /// Order of calling `single` is shuffle with (H1, H2), (H3, H1) and (H2, H3)
+    /// ![Unshuffle steps][unshuffle]
+    #[allow(dead_code)]
+    pub async fn execute_unshuffle<M: Mesh, G: Gateway<M, S>>(
         &mut self,
         ctx: &ProtocolContext<'_, G, S>,
+        permutations: &(Permutation, Permutation),
     ) -> Result<(), BoxError>
     where
         F: Field,
     {
-        *self.input = self.single_shuffle(ctx, Step1).await.unwrap();
-        *self.input = self.single_shuffle(ctx, Step2).await.unwrap();
-        *self.input = self.single_shuffle(ctx, Step3).await.unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Unshuffle, ctx, Step3, permutations)
+            .await
+            .unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Unshuffle, ctx, Step2, permutations)
+            .await
+            .unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Unshuffle, ctx, Step1, permutations)
+            .await
+            .unwrap();
+
+        assert_ne!(
+            &ctx.participant[(self.step_fn)(Step1)],
+            &ctx.participant[(self.step_fn)(Step2)]
+        );
+        assert_ne!(
+            &ctx.participant[(self.step_fn)(Step2)],
+            &ctx.participant[(self.step_fn)(Step3)]
+        );
```

Comment:
Wait... are we in a test? Why are there asserts in production code?

Reply:
Removing it for now